### DISCUSSION TITLE
equivalent of pytorch tensor detach ?

### DISCUSSION THREAD

**Discussion by ninobone-oz** [MrNeRF/LichtFeld-Studio#961] [General]
Very interested in using LichtFeld Studio. 
In looking at porting some pytorch based code, I am not aware how to do the equivalent of pytorch tensor.detach() or .no_grad(). 
I have looked at the implementation of pytorch free tensor code in LichtFeld, doesn't seem to have it as far as I can see.
 
Any ideas?

**Comment by MrNeRF** [comment]
Itdoes not have autograd. All gradients are straight in cuda implemented. So there is no equivalent.

**Reply by ninobone-oz** [reply]
Thanks for the reply.

I want to use a weighted loss function w(r) * |r - gt|, where the weights are per pixel function of the rendered output r (gt is ground truth).
Looks like I need to write my own loss function in cuda. Easiest seems to me to modify:
   ```
 __global__ void fused_l1_kernel(
        const float* __restrict__ img1,
        const float* __restrict__ img2,
        float* __restrict__ grad_out,
        float* __restrict__ partial_sums,
        size_t N,
        float grad_scale) {

 and replace the scalar grad_scale with a tensor containing w(r).
Hope I have got that right! Correct me if I am wrong...
```

**Reply by ninobone-oz** [reply]
the other option I was thinking of was to use an on-the-fly generated mask with compute_photometric_loss_with_mask
